mdata/cache/accnt/flat_accnt.go

```diff
@@ -0,0 +1,310 @@
+	maxSize uint64
+
+	// a last-recently-used implementation that keeps track of all chunks
+	// and which hasn't been used for the longest time. the eviction
+	// function relies on this to know what to evict.
+	lru *LRU
+
+	// whenever a chunk gets evicted a job gets added to this queue. it is
+	// consumed by the chunk cache, which will evict whatever the jobs in
+	// evictQ tell it to
+	evictQ chan *EvictTarget
+
+	// a queue of add and hit events to be processed by the accounting.
+	// each add means data got added to the cache, each hit means data
+	// has been accessed and hence the LRU needs to be updated.
+	eventQ chan *FlatAccntEvent
+
+	stats       *Stats
+	statsTicker *time.Ticker
+	lastPrint   int64
+}
+
+type FlatAccntMet struct {
+	total  uint64
+	chunks map[uint32]uint64
+}
+
+// event types to be used in FlatAccntEvent
+const evnt_complt_met uint8 = 1
+const evnt_miss_met uint8 = 2
+const evnt_part_met uint8 = 3
+const evnt_hit_chnk uint8 = 4
+const evnt_add_chnk uint8 = 5
+
+type FlatAccntEvent struct {
+	t  uint8       // event type
+	pl interface{} // payload
+}
+
+type Stats struct {
+	// metric full hits, all requested chunks were cached
+	complt_met uint32
+
+	// metric complete misses, not a single chunk of the request was cached
+	miss_met uint32
+
+	// metric partial hits, some of the requested chunks were cached
+	part_met uint32
+
+	// new metrics added
+	add_met uint32
+
+	// metrics completely evicted
+	evict_met uint32
+
+	// chunk hits
+	hit_chnk uint32
+
+	// chunk adds
+	add_chnk uint32
+
+	// chunk evictions
+	evict_chnk uint32
+}
+
+// payload to be sent with an add event
+type AddPayload struct {
+	metric string
+	ts     uint32
+	size   uint64
+}
+
+// payload to be sent with a hit event
+type HitPayload struct {
+	metric string
+	ts     uint32
+}
+
+func NewFlatAccnt(maxSize uint64) Accnt {
+	accnt := FlatAccnt{
+		total:       0,
+		metrics:     make(map[string]*FlatAccntMet),
+		maxSize:     maxSize,
+		lru:         NewLRU(),
+		evictQ:      make(chan *EvictTarget, evictQSize),
+		eventQ:      make(chan *FlatAccntEvent, eventQSize),
+		stats:       &Stats{0, 0, 0, 0, 0, 0, 0, 0},
+		lastPrint:   time.Now().UnixNano(),
+		statsTicker: time.NewTicker(time.Second * 10),
+	}
+
+	go accnt.eventLoop()
+	return &accnt
+}
+
+func (a *FlatAccnt) AddChunk(metric string, ts uint32, size uint64) {
+	a.act(evnt_add_chnk, &AddPayload{metric, ts, size})
+}
+
+func (a *FlatAccnt) HitChunk(metric string, ts uint32) {
+	a.act(evnt_hit_chnk, &HitPayload{metric, ts})
+}
+
+func (a *FlatAccnt) MissMetric() {
+	a.act(evnt_miss_met, nil)
+}
+
+func (a *FlatAccnt) PartialMetric() {
+	a.act(evnt_part_met, nil)
+}
+
+func (a *FlatAccnt) CompleteMetric() {
+	a.act(evnt_complt_met, nil)
+}
+
+func (a *FlatAccnt) act(t uint8, payload interface{}) {
+	event := &FlatAccntEvent{
+		t:  t,
+		pl: payload,
+	}
+
+	select {
+	// we never want to block for accounting, rather just let it miss some events and print an error
```

Comment:
IIRC the reason I did that was because the accounting thread is pushing `EvictTarget`s into the `a.evictQ` channel. That channel is consumed by the evict loop, which in turn has to wait for a write lock on the cache while all the workers are using read locks, meaning it might have to wait a bit from time to time. That's a possible reason why accounting can be slightly delayed. Furthermore, assume `evictQ` and `eventQ` both are written in a fashion that blocks if they're full, in that case I'd be afraid of a dead lock scenario where the `evictQ` is blocked because it's waiting for the evict loop to evict a chunk, but the evict loop is waiting for a write lock, but a read lock is held by a worker which is trying to push an `add` event into the `eventQ`. That way I think it could all come to a halt, even though unlikely, still scary.